runtime/bcutil/ROMClassBuilder.cpp

```diff
@@ -1281,18 +1295,48 @@ bool
 ROMClassBuilder::compareROMClassForEquality(U_8 *romClass,bool romClassIsShared,
 		ROMClassWriter *romClassWriter, SRPOffsetTable *srpOffsetTable, SRPKeyProducer *srpKeyProducer,
 		ClassFileOracle *classFileOracle, U_32 modifiers, U_32 extraModifiers, U_32 optionalFlags,
-		ROMClassCreationContext * context)
+		ROMClassCreationContext * context, U_32 romSize, bool isLambda)
 {
-	ComparingCursor compareCursor(_javaVM, srpOffsetTable, srpKeyProducer, classFileOracle, romClass, romClassIsShared, context);
-	romClassWriter->writeROMClass(&compareCursor,
-			&compareCursor,
-			&compareCursor,
-			NULL,
-			NULL,
-			0, modifiers, extraModifiers, optionalFlags,
-			ROMClassWriter::WRITE);
-
-	return compareCursor.isEqual();
+	bool ret = false;
+
+	if (isLambda) {
+		if (sizeof(U_64) < abs((int)(romSize - ((J9ROMClass *)romClass)->romSize))) {
```

Comment:
The HostClassName$$Lambda$ will be the same for both classes at both of these checks because hash key is checked before. So the only difference for the sizes will be the difference between the number of digits of the index number. 

romSize increases by 2 when the number of digits of the index number increases by 1, but romSize has padding and is always multiples of 8. So this romSize check will allow at least 4 but up to 7 difference in the number of digits of the index number. 

classFileSize increases by 1 when the number of digits of the index number increases by 1 and it has no padding so this check will allow exactly 4 digits difference. 

So we would be allowing 4 digits difference in any case which means about 100000 lambda classes under the same host class. (eg. HostClassName$$Lambda$[1-9] can get matched up to HostClassName$$Lambda$99999, HostClassName$$Lambda$[10-99] can get matched up to HostClassName$$Lambda$999999)

Reply:
Can you add some of this additional explanation to the code so we don't lose the detials on why they are different checks?

Reply:
I added some additional explanations to both of the checks. Could you take a look? Thanks.